### PR TITLE
PyGnome initial recipe.

### PR DESCRIPTION
**Pull request by JamesMakela-NOAA**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by ChrisBarker-NOAA** [comment]
I agree to be a maintainer

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:13]: why is linux skipped?

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:26]: let's do >= 3.9

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:39]: I don't think you need all the requirements here -- just the ones that are extra for the tests.

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:60]: we don't need that if we're only doing an import test.

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:30]: this is where the full list of requirements should go -- not in the tests stanza.

**Review by ChrisBarker-NOAA — CHANGES_REQUESTED** [review]
getting there -- putting netcdf et al in host may solve the build issue

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:13]: It is skipped because we didn't verify it on linux.  Don't have an environment for it.  Although our Gitlab runners are on CentOS.  I will enable it just to see if it works.

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:26]: will do...

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:39]: I was failing the import test without the presence of those python packages.

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:60]: Ok, I was testing if we could run the pytests, and I left that in there.  I will remove it.

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:30]: So they should go in the requirements/run?

Comment (ChrisBarker-NOAA) [reply recipes/pygnome/meta.yaml:30]: Exactly -- `requirements/run` is what is need to, well, run the package -- that's what will get installed by conda when you conda install this package.

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:37]: and I think this is the problem -- all those libs it can't find should be in the host requirements, so they'll be there.

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:14]: There's got to be something wrong here -- missing_dso_whitelist  should be only rarely needed in special circumstances !?!?!

Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:14]: Yeah well, this entire configuration file was working just fine locally on my mac.  Warnings?  Yeah a few, but nothing that should have broken the entire thing.

Whatever is causing this has to do with the Azure instances they are running this on.  Tough situation huh?  Something that doesn't even fail locally, but fails catastrophically in the GitHub pipeline.

Comment (ChrisBarker-NOAA) [reply recipes/pygnome/meta.yaml:14]: I suspect it's about the cross-compiling infrastructure -- we're not doing that at all locally, and even if it's not being done here, the system is in place.

Which means we need to figure out "build" vs "host".

Comment (ChrisBarker-NOAA) [reply recipes/pygnome/meta.yaml:14]: you can ping the right group on the PR for help (can't find the list right now) -- and/or ask in the conda community forum:

https://conda.discourse.group/

Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:14]: > I suspect it's about the cross-compiling infrastructure -- we're not doing that at all locally, and even if it's not being done here, the system is in place.
> 
> Which means we need to figure out "build" vs "host".

I am sure that is an important distinction.  However not even the osx64 target is building.  I would like us to first and foremost get that target working since it is what we are familiar with.

File: recipes/pygnome/meta.yaml
Comment (ChrisBarker-NOAA) [inline recipes/pygnome/meta.yaml:28]: Should this be [now win] ?

Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:28]: Maybe.  I don't think it is bothering the windows build step though, so I was keeping it simple.

File: recipes/pygnome/LICENSE.txt
Comment (MementoRC) [inline recipes/pygnome/LICENSE.txt:1]: This file may not be needed as it should be in your source tarball

Comment (JamesMakela-NOAA) [reply recipes/pygnome/LICENSE.txt:1]: That is correct, but how do I get the linter to stop complaining about it?

Comment (MementoRC) [reply recipes/pygnome/LICENSE.txt:1]: Oh, that is odd, it should not complain. A more knowledgeable reviewer should be able to help you further (I am just another recipe maintainer who worked with sckit-build-core)

Comment (JamesMakela-NOAA) [reply recipes/pygnome/LICENSE.txt:1]: Ok, well it isn't hurting anything so I will leave it in to satisfy the linter.

File: recipes/pygnome/meta.yaml
Comment (xhochy) [inline recipes/pygnome/meta.yaml:26]: These should also be in `host`. Maybe you need them here for the `build_platform != target_platform` but really only for that.
```suggestion
    - numpy
    - cython
    - netCDF4
```

File: recipes/pygnome/meta.yaml
Comment (xhochy) [inline recipes/pygnome/meta.yaml:31]: Set `c(xx)_compiler_version` in `conda_build_config` instead. Otherwise, you won't get the right `C(XX)FLAGS) environment variables.
```suggestion
    - {{ compiler('c') }}
    - {{ stdlib('c') }}
    - {{ compiler('cxx') }}
```

File: recipes/pygnome/meta.yaml
Comment (xhochy) [inline recipes/pygnome/meta.yaml:55]: Python packages should always have an explicit `python` in  `run` 
```suggestion
    - python
```

File: recipes/pygnome/meta.yaml
Comment (xhochy) [inline recipes/pygnome/meta.yaml:78]: You're not using it.
```suggestion
```

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml:26]: ok, I made the change.  I put these three items in *both* the build and host sections, and nothing seems to have been broken.

**Comment by ChrisBarker-NOAA** [comment]
@xhochy:

I moved the clang version (15) to conda_build_config -- but now it won't work because apparently somewhere else it's requiring 16 -- ideas?

**Comment by MementoRC** [comment]
Could the recipe need to re re-rendered?

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml]: Ok, the clang 15 requirement is now in conda_build_config.yaml.

File: recipes/pygnome/meta.yaml
Comment (JamesMakela-NOAA) [reply recipes/pygnome/meta.yaml]: It is now removed.

**Comment by ChrisBarker-NOAA** [comment]
@conda-forge/help-python-c r, ready for review!

We've cleaned it up quite a bit from the last one -- hopefully ready to go now!

Thanks!

File: recipes/pygnome/meta.yaml
Comment (xhochy) [reply recipes/pygnome/meta.yaml:26]: In build, they should have a `# [build_platform != target_platform]` comment.

**Review by xhochy — CHANGES_REQUESTED** [review]

**Comment by JamesMakela-NOAA** [comment]
@xhochy:  I added the build_platform comment for numpy, cython, & netCDF4, and I was able to build locally without any failures, but now the checks are failing for all platforms.  The failure in the azure pipeline is:

`2024-08-20T18:13:55.1688580Z   make[2]: CYTHON-NOTFOUND: No such file or directory`

Can you tell me what the purpose was for adding the build_platform comments to these lines.

**Comment by xhochy** [comment]
> Can you tell me what the purpose was for adding the build_platform comments to these lines.

The purpose is to ensure that these dependencies are picked always from the host environment. Only in the cross-compiling case, we will also need to have them present to ensure we can run some binaries natively during the build.

Please re-add it to Cython. If it only works without the comment, there is a deeper issue in dependency discovery with the setup and something else needs to be fixed.

**Comment by ChrisBarker-NOAA** [comment]
>Please re-add it to Cython. If it only works without the comment, there is a deeper issue in dependency discovery with the setup and something else needs to be fixed.

Done -- however, I'm confused -- adding that selector means that Cython will not be a build requirement unless cross-compiling -- so it will never be there at all.

In contrast to the numpy and netCDF4, which are also host and run requirements, so will be there.

Should Cython be a host requirement?

But Cython is not a run-time binary, and needs to be there on the machine running the build, not the target machine -- more like cmake than numpy, which is why we put it in build, without the selector.

so???

**Comment by ChrisBarker-NOAA** [comment]
@xhochy: I put the selector back for cython, and indeed it failed again, 'cause we told it din't need cython unless cross compiling -- so where should we be specifying cython??

It sure seem to me to be a build requirement.

**Comment by ChrisBarker-NOAA** [comment]
cython now in host, but still failures, but only on linux_64. So now I'm really confused.

 I *really* think that cython is a straightforward build requirement:

"""
When both build and host sections are defined, the build section can be thought of as "build tools" - things that run on the native platform, but output results for the target platform.
"""

Cython runs on the native platform, and, indeed, outputs results (C code) that's platform independent. Some projects actually put the generated C code in the sdist, and it can be compiled anywhere. so there's not reason to do anything different when cross compiling.

So why the selector? and it doesn't need to be available at run time (or link time) for the target platform -- so why in host?

Anyway, closer, but not there yet -- we really need this!

**Comment by xhochy** [comment]
> I really think that cython is a straightforward build requirement

No, because cython actually has a Python dependency and should also be using the host Python to produce the correct output. That differentiates it from the normal build/host split. For anything based on Python (in Python packages), it should go primarily in host.

File: recipes/pygnome/meta.yaml
Comment (xhochy) [inline recipes/pygnome/meta.yaml:27]: Same argument as with Cython
```suggestion
    - cython                              # [build_platform != target_platform]
    - scikit-build-core             # [build_platform != target_platform]
```

**Review by xhochy — CHANGES_REQUESTED** [review]

**Comment by ChrisBarker-NOAA** [comment]
we'll see if this works now ....

**Comment by xhochy** [comment]
Can you switch the build to `ninja` so that we can the failing compiler command?

**Comment by ChrisBarker-NOAA** [comment]
> Can you switch the build to ninja

I have no idea what that means, or how to do it. 

I *think* ninja is being used by scikit-build-core -- but that's under the hood as far as I can tell.

Cross-compiling is HARD.

>  Cython ... should also be using the host Python to produce the correct output

That's not how Cython used to work -- has something changed? I thought Cython produced platform-independent C (C++) code. There is *some* dependence on the Python version (i.e. the Python C API), but I don't think there's anything platform dependent in what it generates. In fact, it used to be common practice to include the generated code in an sdist, so it could be built elsewhere without running Cython at all. Has something changed?

Anyway, it all seemed to work with Cython as a regular build dependency -- and now it doesn't :-(

But if ninja can get it work, great, but I have no idea how to do that / what it means?

File: recipes/pygnome/meta.yaml
Comment (xhochy) [reply recipes/pygnome/meta.yaml:31]:
```suggestion
    - ninja
```

**Comment by xhochy** [comment]
> That's not how Cython used to work -- has something changed? I thought Cython produced platform-independent C (C++) code. There is some dependence on the Python version (i.e. the Python C API), but I don't think there's anything platform dependent in what it generates. In fact, it used to be common practice to include the generated code in an sdist, so it could be built elsewhere without running Cython at all. Has something changed?

We basically want to avoid having a Python installation in `build` as long as we aren't cross-compiling.

> But if ninja can get it work, great, but I have no idea how to do that / what it means?

See the suggestion.

**Review by ChrisBarker-NOAA — APPROVED** [review]
maybe there?

**Comment by ChrisBarker-NOAA** [comment]
```
CMake was unable to find a build program corresponding to "Unix Makefiles".  
```
do we need both make and ninja? Or do we need to make out build scripts ninja-based (again, no idea how...)

**Comment by ChrisBarker-NOAA** [comment]
Darn, now getting:

` /Users/runner/Miniforge3/conda-bld/pygnome_1727896937254/work/py_gnome/lib_gnome/Basics.h:15:10: fatal error: 'cstdlib' file not found
  #include <cstdlib>`

That seems pretty basic -- what's up???

**Comment by ChrisBarker-NOAA** [comment]
@conda-forge/help-python-c -- now I"m really confused -- we had it building opn all abut one cross-platform build (I forgot which one) -- but now It's failing on OS-X  with:

```
/Users/runner/Miniforge3/conda-bld/pygnome_1727903416636/work/py_gnome/lib_gnome/Basics.h:15:10: fatal error: 'cstdlib' file not found
  #include <cstdlib>
           ^~~~~~~~~
```
which seems pretty darn basic. Any ideas?

note: we had a lot of struggles with how to get Cython working on the platform builds, but this seems to be something different.

**Comment by xhochy** [comment]
We're running into https://github.com/conda-forge/clangdev-feedstock/issues/310 Seems like `clang=15` is not well supported anymore.

**Comment by h-vetinari** [comment]
> We're running into [conda-forge/clangdev-feedstock#310](https://github.com/conda-forge/clangdev-feedstock/issues/310) Seems like `clang=15` is not well supported anymore.

If you need clang 15, you can just add `libcxx-devel` to your requirements.

**Comment by ChrisBarker-NOAA** [comment]
@h-vetinari : thanks, but `libcxx-devel` is there in the host requirements arlready 

So that's not it :-(

**Comment by h-vetinari** [comment]
The compiler is in the `build:` environment, so `libcxx-devel` should be too.

**Comment by ChrisBarker-NOAA** [comment]
cross your fingers ....

File: recipes/pygnome/meta.yaml
Comment (h-vetinari) [inline recipes/pygnome/meta.yaml:33]: You'll need to limit this to `<19`, because libcxx 19 cannot be parsed by clang 15 anymore.

**Comment by ChrisBarker-NOAA** [comment]
OMG! it's working!   

Thanks all!!!!

don't merge just yet -- I want to make sure we've got this pointed to the latest. and greatest.

**Comment by ChrisBarker-NOAA** [comment]
One question:

This package doesn't work on py < 3.9 and hasn't been tested, may not work, with py3.13, so we have:

```
build:
  skip: True  # [py<39 or py>312 or python_impl == 'pypy']
```

in there.

Should this be handled by pinning python in the requirements instead? If so, in run? or everywhere?

NOTE: I have no idea why we did it that way in the first place -- but I don't want to mess with it unless someone advises me to -- this is building for the first time in months!

**Comment by h-vetinari** [comment]
> Should this be handled by pinning python in the requirements instead?

The skip is fine, though you can drop the pypy part; it's obsolete.

**Comment by ChrisBarker-NOAA** [comment]
OK then -- ready for review / merge!

**Review by xhochy — APPROVED** [review]

**Comment by ChrisBarker-NOAA** [comment]
Thanks all for the support on a long slog!

**Pull request merged**
